get datums using pallas observer

falcucci opened the pull request:
## Content
<!-- Explain the reason for this change, if a feature is added, a bug is fixed, ... -->

Includes:
- get current datums using the new pallas observer;
- brings all compatibilities with the previous cardano-cli format;
- uses the current cardano-cli as a default fallback for the remaining methods;

## Pre-submit checklist

- Branch
  - [x] Tests are provided (if possible)
  - [ ] Crates versions are updated (if relevant)
  - [x] Commit sequence broadly makes sense
  - [x] Key commits have useful messages
- PR
  - [x] No clippy warnings in the CI
  - [x] Self-reviewed the diff
  - [x] Useful pull request description
  - [x] Reviewer requested

## Comments
<!-- Some optional comments about the PR, such as how to run a command, or warnings about usage, .... -->
The automatic fallback mode has been implemented to make sure we progressively rollout the new integration.

falcucci committed:
add pallas current_datums

first integration bring all compatibility

falcucci committed:
add client argument to get_utxo_datums

falcucci committed:
minor changes

falcucci committed:
remove useless code

falcucci committed:
minor changes

falcucci committed:
replace dependencies to our fork

falcucci committed:
removing `println!` call from `cardanoclichainobserver`

the specific changes include:
- removing the `println!` call in `mithril-common/src/chain_observer/cli_observer.rs`
- removing the `minicbor::encode` implementation for `metadatum` in `mithril-common/src/chain_observer/codec.rs`
- removing the `minicbor::encode` implementation for `constr` in `mithril-common/src/chain_observer/codec.rs`
- adding mock utxo data in `mithril-common/src/chain_observer/pallas_observer.rs`

falcucci committed:
update derive attributes for Metadatum struct

falcucci committed:
merge 'main'

falcucci committed:
minor changes

falcucci committed:
removing tag from metadatum

removes the tag field from metadatum and its associated tests.

falcucci committed:
improved error message for inline datum data

falcucci committed:
codec serialization: add correct serializer length

falcucci committed:
improved error handling for Address::from_bech32

falcucci committed:
add Datums docs type

falcucci committed:
filter inline datums

falcucci committed:
Merge branch 'main' into feat/get-datum

falcucci committed:
refactor: `inspect` function to use `vec<u8>` parameter type

- change the function signature of `inspect` to take a `vec<u8>` instead of `anycbor`
- remove the unnecessary decoding of `cbor` in the `inspect` function
- remove the unnecessary creation of `datum` as `anycbor` in the `inspect_datum` function
- remove the unnecessary decoding of `cbor` in the `inspect_datum` function

falcucci committed:
refactor: update comment description for `inspect` function

- change the comment description for the `inspect` function

falcucci committed:
feat: refactor datum inspection

- the `get_datum_bytes` function was added, which returns inline datum bytes from the given `values` instance.
- the implementation of the `inspect_datum` function was modified to use the new `get_datum_bytes` function.

falcucci committed:
apply dry to serialize the external enum

falcucci committed:
style: improve code clarity, testing, and github actions scheduling

- remove unneeded comment about derive attributes in chain_observer model.

falcucci committed:
chore: update pallas dependencies and repositories

- switched to fetching pallas libraries directly from git repository instead of local paths
- updated the source of `pallas-addresses`, `pallas-codec`, `pallas-crypto`, `pallas-network`, and `pallas-primitives` packages in the cargo.lock file and mithril-common's cargo.toml file

falcucci commented:
@scarmuega ready for review,

we also need a new pallas release to bump the new version with the get utxo by address state query in order to fix the [check](https://github.com/txpipe/mithril/actions/runs/7290537684/job/19867554439?pr=1) ci step :)

falcucci committed:
refactor: improve serialization process of `datum` type

- import and utilize `tocanonicaljson` in `chain_observer/codec.rs` for conversion of `datum` type to json, simplifying the original implementation
- add a `serialize` derivation for the `datum` struct in `chain_observer/model.rs`
- update the `serialize_datum` function in `chain_observer/pallas_observer.rs` to use the new `to_json()` method for serialization.

falcucci committed:
refactor: code and improve testing accuracy

- removed `bigint` and `constr` from the `conway` module in the `chain_observer`'s `codec.rs` file

falcucci committed:
refactor: utxo handling and streamline byte decoding

- add a new function `inspect` in `codec.rs` to decode `r` instances from bytes
- remove the `inspect` function from `pallas_observer.rs` and use the `inspect` function from `codec.rs` instead for decoding bytes
- refactor replacement of `values` with `utxo` in `get_datum_bytes`, `inspect_datum`, and `serialize_datum` functions
- change `map_datums` function to process `transaction.utxo` instead of `utxo.utxo`

falcucci committed:
refactor: improve test robustness and broaden platform support

- refactor the 'values' struct initialization to include type conversions for the 'address' field in the testing module of the 'pallas_observer.rs' file.

falcucci commented:
@scarmuega the canonical trait worked like a charm!

all the requests was addressed and is ready for a new review.

thanks

scarmuega commented on mithril-common/src/chain_observer/codec.rs:
```diff
@@ -0,0 +1,24 @@
+use pallas_primitives::conway::PlutusData;
+use pallas_primitives::ToCanonicalJson;
+
+use super::model::Datum;
+
+impl<'a, C> minicbor::Decode<'a, C> for Datum {
+    fn decode(d: &mut minicbor::Decoder<'a>, ctx: &mut C) -> Result<Self, minicbor::decode::Error> {
+        PlutusData::decode(d, ctx).map(Datum)
+    }
+}
+
+impl ToCanonicalJson for Datum {
+    fn to_json(&self) -> serde_json::Value {
+        self.0.to_json()
+    }
+}
+
+/// Inspects the given bytes and returns a decoded `R` instance.
+pub fn inspect<R>(inner: Vec<u8>) -> R
+where
+    for<'b> R: pallas_codec::minicbor::Decode<'b, ()>,
+{
+    minicbor::decode(&inner).unwrap()
+}
```
This is just short boilerplate code, no need to put it on a different mod. I would inline it inside cli_observer.

scarmuega commented on mithril-common/src/chain_observer/pallas_observer.rs:
```diff
@@ -78,6 +84,81 @@ impl PallasChainObserver {
+        statequery
+            .acquire(None)
+            .await
+            .map_err(|err| anyhow!(err))
```
is `map_err(|err| anyhow!(err))` necessary? doesn't anyhow already transform the error type automatically?

falcucci replied:
yes but on mithril they use as much as possible `with_context` to provide additional context in the error type, so due to that the `map_err` should be needed AFAIK

falcucci committed:
refactor: codec functionality into model.rs

- deleted the `codec.rs` file from `mithril-common/src/chain_observer/`
- removed the reference to `codec` in `mod.rs`
- moved the `decode` and `tocanonicaljson` implementations for `datum`, and the `inspect` function, from `codec.rs` to `model.rs`
- adjusted the imports in `pallas_observer.rs` to reflect the moved functionality from `codec.rs` to `model.rs`

falcucci committed:
Merge branch 'main' into feat/get-datum

falcucci committed:
chore: update dependencies and refactor chain observer

- added `minicbor` package version `0.20.0` to dependencies.
- updated `minicbor` package from version `0.19.1` to `0.19.2`.
- updated `pallas-addresses`, `pallas-codec`, `pallas-crypto`, `pallas-network`, `pallas-primitives` packages from version `0.20.0` to `0.21.0` and changed the source.
- updated packages `pallas-addresses`, `pallas-codec`, `pallas-network`, `pallas-primitives`, `pallas-crypto` in `mithril-common/cargo.toml` from using branch to using exact version number `0.21.0`.
- in `mithril-common/src/chain_observer/model.rs`, replaced `minicbor::decode` with `decode` from `pallas_codec::minicbor`.
- in `mithril-common/src/chain_observer/pallas_observer.rs`, replaced calls to `drop(client.plexer_handle)` with `client.abort().await`.

falcucci committed:
refactor: update plutusdata import from conway to alonzo

- replace `conway::plutusdata` with `alonzo::plutusdata` in the import statement.

falcucci committed:
chore: update dependencies and adjust mithril-common features

- updated `minicbor` dependency from version `0.19.1` to version `0.20.0`
- removed an older version of `minicbor` from dependencies
- made `pallas-addresses`, `pallas-codec`, and `pallas-primitives` optional dependencies in `mithril-common/cargo.toml`
- added optional dependencies of `mithril-common` to `fs` feature: `pallas-addresses`, `pallas-primitives`, and `pallas-codec`

falcucci committed:
chore: refactor dependencies in `mithril-common` package

- replace the dependency in the `mithril-common` package from `pallas-network`, `pallas-addresses`, `pallas-primitives`, and `pallas-codec` to only `pallas-codec`.

falcucci committed:
minor changes

falcucci committed:
refactor: mock utxo function name and update test calls

- rename `get_utxo_mock()` function to `get_fake_utxo_by_address()`
- update the function calls accordingly within the testing scope

falcucci committed:
refactor: use full path for pallas_codec::minicbor calls

- remove direct import of `decode` and `decoder` from `pallas_codec::minicbor`
- call `decode` implementation of `datum` via full path `pallas_codec::minicbor::decode`
- call function `decode` with full path `pallas_codec::minicbor::decode` in the `inspect` function

falcucci committed:
refactor: decoding using `minicbor::decode`

- import `minicbor::decode` in `chain_observer/model.rs`
- refactor `datum` implementation to use `minicbor::decode` instead of `pallas_codec::minicbor::decode`
- changed `inspect` function to decode with `minicbor::decode` instead of `pallas_codec::minicbor::decode`

falcucci committed:
feat: implement feature-specific code with `cfg_fs` macro

- implement a new macro named `cfg_fs` for enabling feature specific code
- move `use pallas_primitives::{alonzo::plutusdata, tocanonicaljson};` statement inside `cfg_fs` macro block in `model.rs`

falcucci committed:
refactor: import and implementation placement in `cfg_fs` block

- the `pallas_primitives` import gets moved inside the `cfg_fs` block.
- the `tocanonicaljson` implementation for `datum` gets moved within the same `cfg_fs` block.

falcucci committed:
refactor: datum decoding and update tests

- remove the import of `minicbor::decode` in `chain_observer/model.rs`.
- adjust the specification of the `datum` decoding function to reflect the removal of the `minicbor::decode` import.
- add a closing brace that appears to define a new code block, possibly a struct or a function.

falcucci committed:
refactor: `deserialize` import location in `cfg_fs!` macro

- move the `deserialize` import from the root to the `cfg_fs!` macro.
- there were no other significant changes.

falcucci committed:
style: "standardize case in pallaschainobserver error messages"

- correct the case in error messages from "pallaschainobserver failed to acquire statequery" to "pallaschainobserver failed to acquire statequery"
- adjust the case in error message from "pallaschainobserver failed to get current era" to "pallaschainobserver failed to get current era"
- alter the case in error message from "pallaschainobserver failed to get block epoch number" to "pallaschainobserver failed to get block epoch number"

falcucci committed:
refactor: optimize inline_datum handling in pallaschainobserver

- refactor the method of filtering and mapping inline_datum in `pallaschainobserver` to use `filter_map` for efficiency.

falcucci committed:
refactor: improve error handling in model and observer files

- add `anyhow::context` in `model.rs`
- change the function `inspect` to `try_inspect`, which now returns a `stdresult`, modifies the decoding process and includes error context
- replace usages of `inspect` with `try_inspect` in `pallas_observer.rs` and add failure messages for possible decoding errors
- tweak the filter_map function in `get_datums` for readability

falcucci committed:
refactor: improve error handling in pallaschainobserver functions

- update the return types for `get_datum_bytes`, `inspect_datum`, `serialize_datum`, and `map_datums` functions in the `pallaschainobserver` implementation to return a `stdresult`.
- remove calls to expect in favour of handling potential errors.
- modify `serialize_datum` function to include better error context when serialization fails.
- allow error propagation in `map_datums` and `get_utxo_by_address` function calls.
- the methods mentioned changes make the code more robust by improving error management and providing better debugging context.

falcucci committed:
refactor: refine error handling and testing mechanisms

- removed an unnecessary ok wrap from the minicbor decode function's error handling in the model.rs file.

falcucci committed:
refactor: `try_inspect` usage in `pallaschainobserver`

- refactor the `try_inspect` function call in `get_datum_bytes` method of `pallaschainobserver` in the pallas observer implementation to provide explicit datatype for inspection.

falcucci committed:
feat: refactor datum retrieval methods in pallaschainobserver

- import bytes and tagwrap from `pallas_codec::utils`
- replace the existing implementation of `get_datum_bytes` method. now it retrieves the inline datum tag, wraps it with `cborwrap` and returns it as a vector
- add new `get_datum_tag` method to the `pallaschainobserver` structure. this new method extracts and returns the inline datum tag from the provided `values` instance

scarmuega commented on mithril-common/src/chain_observer/model.rs:
```diff
@@ -1,20 +1,49 @@
-#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
-#[serde(rename_all = "lowercase")]
-pub struct Datum(pub PlutusData);
+cfg_fs! {
```
Why is this macro require? does it have to do with fs = filesystem? I don't see any access to the file system in this block.

scarmuega commented on mithril-common/src/chain_observer/pallas_observer.rs:
```diff
@@ -69,51 +68,66 @@ impl PallasChainObserver {
             .await
             .map_err(|err| anyhow!(err))
-            .with_context(|| "PallasChainObserver Failed to acquire statequery")?;
+            .with_context(|| "PallasChainObserver failed to acquire statequery")?;
```
I think that for static message errors is more efficient to use `context(str)` instead of `with_context(fn)`

falcucci replied:
the previous implementation was using `context` but it was requested to use `with_context` on non test code.

It seems to be an ADR from the mithril team, here: https://mithril.network/doc/adr/6

scarmuega commented on mithril-common/src/lib.rs:
```diff
@@ -10,6 +10,16 @@
 //! - A [certificate chain] used to validate the Certificate Chain created by an aggregator
 //! - The [entities] used by, and exchanged between, the aggregator, signers and client.
 
+macro_rules! cfg_fs {
+    ($($item:item)*) => {
+        $(
+            #[cfg(feature = "fs")]
+            #[cfg_attr(docsrs, doc(cfg(feature = "fs")))]
+            $item
+        )*
+    }
+}
+
```
why is this required?

falcucci replied:
due to wasm [build](https://github.com/input-output-hk/mithril/actions/runs/7422372066/job/20197520073?pr=1403) errors, the mithril team suggested a conditional compilation macro.

more details here: https://github.com/input-output-hk/mithril/pull/1403/files/acccd103df4ad21cd126de6e14003135da7105b0#r1442959628

falcucci committed:
style: refactor error messaging and numerical tolerance adjustments

- corrected the capitalization in error messages to be consistent in `pallaschainobserver`. specifically, changed `failed` to `failed` in various context messages.

falcucci committed:
refactor: code and improve test and api reliability

- refactor import structure of `pallas_network` in `pallas_observer.rs`

falcucci committed:
refactor: refactor to use `minicbor` decode trait directly

- import `decode, decoder, decode` functions from `minicbor` instead of using them with full path.
- refactor `datum` impl block to use `decode` trait from `minicbor` in a more straight forward way.
- change `try_inspect` function to use `decode` trait and `decode` function directly from `minicbor`.

falcucci committed:
Merge branch 'main' into feat/get-datum

falcucci committed:
chore: integrate minicbor as an optional feature in mithril-common

- change `minicbor` to be an optional feature in `mithril-common/cargo.toml`
- add `minicbor` to the `fs` feature list.

falcucci committed:
Merge branch 'main' into feat/get-datum

falcucci committed:
Merge branch 'main' into feat/get-datum

falcucci committed:
merge 'main'

falcucci committed:
refactor: datum retrieval in pallaschainobserver for thales supported era

- remove the function `get_datum_bytes` from `pallaschainobserver`
- simplify obtaining the datum in `inspect_datum` by directly using `get_datum_tag`
- update the test `hex_datum` value and corresponding assertion in the test module

falcucci committed:
refactor: usage of anyhow::context in cfg_fs block

- remove the `use anyhow::context;` from the start of the file.
- add `use anyhow::context;` within the `cfg_fs!` block.

falcucci committed:
merge 'main'

falcucci committed:
chore: switch pallas dependencies to fix branch and update usage

- change source of `pallas-addresses`, `pallas-codec`, `pallas-network`, `pallas-primitives`, `pallas-traverse`, and `pallas-crypto` in `cargo.lock` and `mithril-common/cargo.toml` from crates.io to a specific branch `fix/legacy-transaction-utxo-by-address` on repository `https://github.com/falcucci/pallas.git`
- replace `values` type with `postalonsotransactionoutput` as an argument in `pallaschainobserver` methods including `get_datum_tag`, `inspect_datum`, and `serialize_datum` in `mithril-common/src/chain_observer/pallas_observer.rs`
- update filter logic in `serialize_datum` to match `transactionoutput::map`
- update test data to match the new `transactionoutput` type changes

falcucci committed:
chore: upgrade package versions and update utxo iteration

- update `inventory` from version `0.3.14` to `0.3.15` with new checksum.
- update multiple `pallas` packages (addresses, codec, crypto, network, primitives, traverse) from version `0.21.0` to `0.22.0` with updated git source reference.
- update `snow` from version `0.9.5` to `0.9.6` with new checksum.
- in `pallas_observer.rs`, replace `transactionoutput::map` with `transactionoutput::current` for utxo iteration and test values instantiation.

falcucci committed:
Merge branch 'main' into feat/get-datum

falcucci committed:
refactor: improve data serialization and testing accuracy

- modify the pallaschainobserver implementation to serialize the output datum inline if present, instead of unconditionally.

falcucci committed:
merge 'main'

falcucci committed:
merge 'main'

falcucci committed:
chore: update pallas dependencies to version 0.23.0

- updated `pallas-addresses`, `pallas-codec`, `pallas-network`, `pallas-primitives`, `pallas-traverse`, and `pallas-crypto` dependencies from git to version `0.23.0`.

falcucci committed:
refactor: `mithril-common` for macro removal

- removed the `cfg_fs` macro from `lib.rs` in the `mithril-common` module.

falcucci committed:
merge 'main'

falcucci committed:
bump up version of `mithril-common` from `0.3.0` to `0.3.1`

- bump up version of `mithril-common` from `0.3.0` to `0.3.1`.
